File: LeetCode/498.Diagonal_Traverse/498.Diagonal_Traverse.py
#!/usr/bin/env python3


# bfs
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        # bfs
        # (i,j) -> (i+1, j) if i == 0, (i, j+1)

        m, n = len(mat), len(mat[0])
        q = deque([])
        q.append((0, 0))

        res = []
        even = False
        while q:
            size = len(q)
            tmp = []
            for _ in range(size):
                i, j = q.popleft()
                tmp.append(mat[i][j])
                if i + 1 < m:
                    q.append((i + 1, j))
                if i == 0 and j + 1 < n:
                    q.append((i, j + 1))
            if even:
                tmp.reverse()
            res += tmp
            even = not even

        return res


# simulate the whole process
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        m, n = len(mat), len(mat[0])
        k = (m - 1) + (n - 1)

        res = []
        for d in range(k + 1):
            even = (d + 1) % 2 == 0

            # start point
            row = 0 if d < n else d - (n - 1)
            col = d if d < n else n - 1

            # one pass
            tmp = []
            while row < m and col >= 0:
                tmp.append(mat[row][col])
                row += 1
                col -= 1

            if not even:
                tmp.reverse()

            res += tmp

        return res

File: LeetCode/498.Diagonal_Traverse/498.Diagonal_Traverse.swift
class Solution {
    func findDiagonalOrder(_ mat: [[Int]]) -> [Int] {
        // key = i + j
        let m = mat.count, n = mat[0].count
        var diag = Array(repeating: [Int](), count: m + n - 1)
        for i in 0 ..< m {
            for j in 0 ..< n {
                diag[i + j].append(mat[i][j])
            }
        }

        var res: [Int] = []
        for (i, lst) in diag.enumerated() {
            if i & 1 == 1 {
                res.append(contentsOf: lst)
            } else {
                res.append(contentsOf: Array(lst.reversed()))
            }
        }
        return res
    }
}
